File: leetcode/numbers/easy/java_solutions/GuessNumberHigherLower_374.java
package leetcode.numbers.easy.java_solutions;

public class GuessNumberHigherLower_374 {

    private static class GuessGame {

        /**
         * Forward declaration of guess API.
         *
         * @param num your guess
         * @return -1 if num is higher than the picked number
         * 1 if num is lower than the picked number
         * otherwise return 0
         * int guess(int num);
         */
        protected int guess(int num) {
            return -1;
        }
    }

    static class Solution extends GuessGame {

        public int guessNumber(int n) {
            int min = 1;
            int max = n;

            while (true) {
                int middle = min + (max - min) / 2;
                int result = guess(middle);
                if (result == 0) {
                    return middle;
                } else if (result > 0) {
                    min = middle + 1;
                } else {
                    max = middle - 1;
                }
            }
        }
    }
}

File: leetcode/numbers/easy/golang_solutions/guess_number_higher_lower_374.go
package golang_solutions

/**
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return 	     -1 if num is higher than the picked number
 *			      1 if num is lower than the picked number
 *               otherwise return 0
 * func guess(num int) int;
 */

func guess(num int) int {
	return -1
}

func guessNumber(n int) int {
	min, max := 1, n
	for {
		middle := min + (max-min)/2

		temp := guess(middle)

		if temp == 0 {
			return middle
		} else if temp > 0 {
			min = middle + 1
		} else {
			max = middle - 1
		}
	}
}
